src/main/java/com/github/firmwehr/gentle/lexer/Lexer.java

```diff
@@ -1,100 +1,321 @@
 package com.github.firmwehr.gentle.lexer;
 
-import com.github.firmwehr.gentle.lexer.tokens.Token;
+import com.github.firmwehr.gentle.parser.tokens.CommentToken;
+import com.github.firmwehr.gentle.parser.tokens.EofToken;
+import com.github.firmwehr.gentle.parser.tokens.IdentToken;
+import com.github.firmwehr.gentle.parser.tokens.IntegerLiteralToken;
+import com.github.firmwehr.gentle.parser.tokens.Keyword;
+import com.github.firmwehr.gentle.parser.tokens.KeywordToken;
+import com.github.firmwehr.gentle.parser.tokens.Operator;
+import com.github.firmwehr.gentle.parser.tokens.OperatorToken;
+import com.github.firmwehr.gentle.parser.tokens.Token;
+import com.github.firmwehr.gentle.parser.tokens.WhitespaceToken;
 import com.github.firmwehr.gentle.source.Source;
-import com.google.common.base.Preconditions;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import com.github.firmwehr.gentle.source.SourceSpan;
 
-import java.util.Set;
-import java.util.function.Predicate;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
 
-/**
- * Primary class for lexing strings.
- */
 public class Lexer {
 
-	private static final Logger LOGGER = LoggerFactory.getLogger(Lexer.class);
+	private final StringReader reader;
+	private final boolean onlyRelevantTokens;
 
-	private final Predicate<TokenType> tokenFilter;
-	private final Source source;
+	public Lexer(Source source, boolean onlyRelevantTokens) {
+		this.reader = new StringReader(source);
+		this.onlyRelevantTokens = onlyRelevantTokens;
+	}
 
-	private LexReader reader;
+	public List<Token> lex() throws LexerException {
+		List<Token> tokens = new ArrayList<>();
 
-	/**
-	 * Constructs a new lexer that will generate tokens from the given source.
-	 *
-	 * @param source The source file to read.
-	 */
-	public Lexer(Source source) {
-		this(source, tokenType -> true);
+		while (reader.canRead()) {
+			if (isDigit(reader.peek())) {
+				tokens.add(readNumber());
+				continue;
+			}
+			if (isIdentifierStart(reader.peek())) {
+				tokens.add(readIdentOrKeyword());
+				continue;
+			}
+			if (Character.isWhitespace(reader.peek())) {
+				int start = reader.getPosition();
+				reader.readWhitespace();
+				if (!onlyRelevantTokens) {
+					tokens.add(new WhitespaceToken(new SourceSpan(start, reader.getPosition())));
```

Comment:
Aren't all these `SourceSpan` end offsets one too large? The end offset should point to the last character of the token, according to the `SourceSpan` documentation. (We can probably just change the documentation now that everything works on offsets)